### PR TITLE
[WIP] Attempt to retry pending requests on actor restart

### PR DESCRIPTION
**Pull request by apaszke**
Previously any actor failure would cause all outstanding actor tasks to
get marked as failed, leading to an exception in the user program.
This patch changes the behavior such that as long as the actor has
remaining reconstructions, all tasks which haven't been completed previously
will get resubmitted to the new instance automatically.

The new semantics allow to handle failures in a more transparent way, but
also requiresthe user to make sure that the actor will always restart in a
state in which the repeated calls are still valid. If this is not the desired
behavior the workaround is to disallow automatic reconstruction and implement
manual handling of actor failures in the user program.

## Related issue numbers

#6670, #6773

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/common/task/task_spec.h [hunk 1]
```diff
@@ -153,6 +153,8 @@ class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
 
   uint64_t ActorCounter() const;
 
+  void SetActorCounter(uint64_t actor_counter);
```
Comment (apaszke) [inline src/ray/common/task/task_spec.h]: I have mixed feelings about this change, because so far this class seems to only have immutable methods, but I need to update the sequence number before I retry submitting the task. An alternative would be to try to recreate the spec from scratch, but (1) it might be quite verbose and difficult to do from within the `TaskManager` and (2) it's likely to be slower.

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 2]
```diff
@@ -48,14 +48,12 @@ class CoreWorkerDirectActorTaskSubmitter {
...
-  /// \return Status::Invalid if the task is not yet supported.
-  Status SubmitTask(TaskSpecification task_spec);
+  void SubmitTask(TaskSpecification task_spec);
 
   /// Tell this actor to exit immediately.
   ///
   /// \param[in] actor_id The actor_id of the actor to kill.
-  /// \return Status::Invalid if the actor could not be killed.
-  Status KillActor(const ActorID &actor_id);
+  void KillActor(const ActorID &actor_id);
```
Comment (apaszke) [inline src/ray/core_worker/transport/direct_actor_transport.h]: Those methods never failed so I figured it's quite useless to return `Status`, but I can change it back if you think this violates some project conventions.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 3]
```diff
@@ -83,40 +73,26 @@ void CoreWorkerDirectActorTaskSubmitter::ConnectActor(const ActorID &actor_id,
...
+  // any tasks submitted after the actor death.
+}
+
+void CoreWorkerDirectActorTaskSubmitter::DisconnectActorUnprotected(const ActorID &actor_id) {
+  rpc_clients_.erase(actor_id);
+  worker_ids_.erase(actor_id);
+  pending_force_kills_.erase(actor_id);
+  pending_requests_.erase(actor_id);
+  next_send_position_.erase(actor_id);
+  next_send_position_to_assign_.erase(actor_id);
```
Comment (apaszke) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: It might be a good idea to have a single map from `ActorID` to a struct holding all of those 🤔

File: src/ray/core_worker/task_manager.h [hunk 4]
```diff
@@ -93,13 +121,39 @@ class TaskManager : public TaskFinisherInterface {
...
 
   /// Return the number of pending tasks.
   int NumPendingTasks() const { return pending_tasks_.size(); }
 
+  /// Return the number of tasks pending for a specific actor.
+  int NumPendingActorTasks(const ActorID &actor_id) {
+    auto it = pending_actor_tasks_.find(actor_id);
+    RAY_CHECK(it != pending_actor_tasks_.end());
+    return it->second.size();
+  }
```
Comment (apaszke) [inline src/ray/core_worker/task_manager.h]: Not sure if there's a policy around inline methods. Can move this to `.cc`, but it seems like a good candidate for inlining.

**Comment by stephanie-wang** [comment]
Hmm I didn't quite understand how killing the actor would restart its tasks. Is the idea that eventually the notification will reach the `TaskManager`, which will then retry the tasks?

I was actually thinking we could just reuse the existing retry functionality in the `TaskManager` that just calls `SubmitTask` again. One way to do it would be to have the `CoreWorkerDirectActorSubmitter` erase its RPC client upon receiving an actor task error. The `TaskManager` would immediately retry the actor task, the same as if it were a normal task, but it will just get queued since there is no known actor location. This wouldn't handle transient errors where the actor is actually still alive, but neither does the approach of killing the actors on any error.

It seems like the main complication would be in making sure that we reset the counters properly. I think what you have to reset the `ActorHandle`'s counter looks good, but for the above approach, I believe we'd also have to make sure to reset the `CoreWorkerDirectActorSubmitter`'s `next_send_position_` correctly.

File: src/ray/core_worker/task_manager.h (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/core_worker/task_manager.h]: We don't have a specific convention, but this seems fine. One weird thing, though, is that I think the thread safety analysis wasn't working for header files, but I could be wrong (@edoakes?).

File: src/ray/common/task/task_spec.h (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/common/task/task_spec.h]: Yeah, I think it's okay to reset the actor counter. Normally we do try to keep the `TaskSpecification` const, but there are special cases where we do actually want to modify it. Another example of this is when we modify the task args to inline any small objects that are now available.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 5]
```diff
@@ -155,11 +131,13 @@ void CoreWorkerDirectActorTaskSubmitter::PushActorTask(
   rpc::Address addr(client.Addr());
   RAY_CHECK_OK(client.PushActorTask(
       std::move(request),
-      [this, addr, task_id](Status status, const rpc::PushTaskReply &reply) {
+      [this, addr, actor_id, task_id](Status status, const rpc::PushTaskReply &reply) {
         if (!status.ok()) {
-          task_finisher_->PendingTaskFailed(task_id, rpc::ErrorType::ACTOR_DIED, &status);
+          // Try to restart the task by killing the actor. This is not a great way to deal with
+          // failures, but a more robust solution would require extra work (see #6773).
+          KillActor(actor_id);
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: We need to be a bit careful here because this callback could actually fire after the actor has already restarted.

Comment (apaszke) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Yeah I'm definitely not happy with how this works right now, but I'm also not sure how to accurately determine the state of the actor right now without suffering from all those races. This may cause the actor to get killed unnecessarily (and also waste some reconstructions), but it will do the job of re-delivering the requests, so I guess it's still better than what happened before 🤷‍♂

A proper fix would probably involve an attempt to resend this particular message, but that's not very easy because the client is super stateful (e.g. it is entirely in charge of sequence numbers and doesn't expose them to anyone else, so right now there's no easy way to resend a message). Do you have any other suggesions?

**Comment by apaszke** [comment]
> Hmm I didn't quite understand how killing the actor would restart its tasks. Is the idea that eventually the notification will reach the TaskManager, which will then retry the tasks?

Exactly. Killing the actor will cause the raylet (or whatever is in charge of reconstructing the actor) to initialize it once more, sending a notification to `CoreWorker` about the state change, which in turn will eventually trigger `task_manager->ResubmitTasks(actor_id)` once the actor gets up.

> I was actually thinking we could just reuse the existing retry functionality in the TaskManager that just calls SubmitTask again. One way to do it would be to have the CoreWorkerDirectActorSubmitter erase its RPC client upon receiving an actor task error. The TaskManager would immediately retry the actor task, the same as if it were a normal task, but it will just get queued since there is no known actor location. This wouldn't handle transient errors where the actor is actually still alive, but neither does the approach of killing the actors on any error.

I'm not sure if I understand your proposal. It seems like it would deadlock in the case of a transient error when one task push fails, but the actor doesn't crash, because it would never trigger `CoreWorkerDirectActorSubmitter::ConnectActor` again causing the messages to wait in the queue indefinitely. Killing the actor forces the connection to be recreated in this case.

File: src/ray/core_worker/task_manager.cc [hunk 6]
```diff
@@ -10,11 +10,10 @@ const int64_t kTaskFailureThrottlingThreshold = 50;
 // Throttle task failure logs to once this interval.
 const int64_t kTaskFailureLoggingFrequencyMillis = 5000;
 
-void TaskManager::AddPendingTask(const TaskID &caller_id,
+void TaskManager::AddPendingTaskImpl(const TaskID &caller_id,
```
Comment (edoakes) [inline src/ray/core_worker/task_manager.cc]: I think in other places in the code (e.g., in `reference_count.h`) we call similar methods `XyzInternal`. No strong opinion but would be good to be consistent.

File: src/ray/core_worker/task_manager.cc [hunk 7]
```diff
@@ -52,6 +51,24 @@ void TaskManager::AddPendingTask(const TaskID &caller_id,
...
+  AddPendingTaskImpl(caller_id, caller_address, spec, max_retries);
+}
+
+void TaskManager::AddPendingActorTask(const TaskID &caller_id,
+                                 const rpc::Address &caller_address,
+                                 const TaskSpecification &spec) {
+  RAY_CHECK(spec.IsActorTask());
+  absl::MutexLock lock(&mu_);
+  AddPendingTaskImpl(caller_id, caller_address, spec, /*max_retries=*/0);
+  pending_actor_tasks_[spec.ActorId()].push_back(spec.TaskId());
```
Comment (edoakes) [inline src/ray/core_worker/task_manager.cc]: Should we be checking that `spec.ActorId()` is in the map here? If we're sure that it must be, would probably be worth a comment. Might also be nice to have an explicit `RAY_CHECK` to give a better error message/make it clearer.

File: src/ray/core_worker/task_manager.cc [hunk 8]
```diff
@@ -52,6 +51,24 @@ void TaskManager::AddPendingTask(const TaskID &caller_id,
...
+  actor_send_task_callback_(spec);
```
Comment (edoakes) [inline src/ray/core_worker/task_manager.cc]: Need to be careful about calling this while holding `mu_` (both for possible deadlock and performance)

File: src/ray/core_worker/task_manager.cc [hunk 9]
```diff
@@ -118,6 +135,34 @@ void TaskManager::CompletePendingTask(const TaskID &task_id,
   ShutdownIfNeeded();
 }
 
+void TaskManager::CompletePendingTask(const TaskID &task_id,
+                                      const rpc::PushTaskReply &reply,
+                                      const rpc::Address &worker_addr) {
+  // Make sure that task_id doesn't correspond to an actor task.
```
Comment (edoakes) [inline src/ray/core_worker/task_manager.cc]: This seems like a very heavyweight check - is it necessary/is there some way to do it without grabbing & releasing the lock an additional time?

File: src/ray/core_worker/task_manager.h (hunk 4 above)
Comment (edoakes) [reply src/ray/core_worker/task_manager.h]: That's right - we should avoid having code in header files for any class with thread annotations.

**Review by edoakes — COMMENTED** [review]
A few small comments but will hold off on more until you restructure the code as discussed.

File: src/ray/core_worker/transport/direct_actor_transport.cc (hunk 5 above)
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Long-term, it probably makes sense to version the messages and clients, so we should only kill the actor if its version matches the one sent with the task. For now, it's probably okay for this to be a bit inefficient.

File: src/ray/core_worker/task_manager.cc [hunk 10]
```diff
@@ -186,6 +231,31 @@ void TaskManager::ShutdownIfNeeded() {
...
+    absl::MutexLock lock(&mu_);
+    actor_tasks = std::move(pending_actor_tasks_[actor_id]);
+    pending_actor_tasks_.erase(actor_id);
+  }
+  for (auto& task_id : actor_tasks) {
+    PendingTaskFailed(task_id, rpc::ErrorType::ACTOR_DIED);
+  }
+}
+
+void TaskManager::ResubmitActorTasks(const ActorID &actor_id) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/task_manager.cc]: Btw, another thought I had is that it probably wouldn't work to just resubmit everything anyway, since some tasks may be buffered in the `CoreWorkerDirectActorSubmitter`, so we could end up with duplicate task submissions. Maybe you've already accounted for that, but just another thing to be aware of.

**Comment by stephanie-wang** [comment]
After some discussion offline, I think the plan is now this:
1. Move the sequence numbers out of the `CoreWorkerDirectActorSubmitter` and into the `CoreWorker`. The only responsibility of the `CoreWorkerDirectActorSubmitter` is to send actor tasks in the order of the assigned sequence number.
2. On an actor task failure, the  `CoreWorkerDirectActorSubmitter` should clear its RPC client to prevent further tasks from getting sent and start a timer. If no message is heard about the actor within the timer, then all tasks are failed (optional whether to include this in the PR). Also, give the failed task back to the `TaskManager`, who will decide whether it should be resubmitted or not.
3. If the actor is restarted, then the `CoreWorkerDirectActorSubmitter`'s last sent sequence number is reset to the total number of tasks that completed successfully before, across all previous versions of the actor.

Does this sound right, @apaszke?

File: src/ray/core_worker/task_manager.cc (hunk 10 above)
Comment (apaszke) [reply src/ray/core_worker/task_manager.cc]: Yeah I explicitly flush the submitter every time a connection is dropped (which didn't happen previously).

**Comment by apaszke** [comment]
@stephanie-wang Thanks for writing that up! That does sound correct, although IIRC we did say that in case of a timeout the `CoreWorkerDirectActorSubmitter` should kill the actor to trigger another retry instead of failing the tasks. The tasks only fail when the actor dies.

**Comment by stephanie-wang** [comment]
> @stephanie-wang Thanks for writing that up! That does sound correct, although IIRC we did say that in case of a timeout the `CoreWorkerDirectActorSubmitter` should kill the actor to trigger another retry instead of failing the tasks. The tasks only fail when the actor dies (in the sense that it has used up all its reconstructions).

Ah yes, thanks for the clarification!

**Comment by stephanie-wang** [comment]
Okay, after thinking about this some more, I think I finally understand why the approach I was suggesting makes the state complicated to deal with. I forgot that you also have to `SetActorCounter` for all of the tasks that need to get resubmitted. Would it help if we extend what we decided on with the following (sorry if this was already clear to you, but just trying to clarify for myself)?
1. Add a version to the `ActorHandle` and the `TaskSpecification` that indicates which version of the actor the task is for. Track the version in the `TaskManager` and the `CoreWorkerDirectActorSubmitter` (the latter should associate the version with the RPC client).
2. In the `CoreWorkerDirectActorSubmitter`, if a task fails, also fail all pending but unsent tasks and increment the actor version. The `TaskManager` will perform the timeout by waiting for the next actor location to appear, as we described. If it reappears, then resubmit the tasks with the `TaskManager::ResubmitActorTasks`, like you've implemented already.
3. In the `CoreWorkerDirectActorSubmitter`, if a task with an old version is submitted, fail it immediately and let the `TaskManager` retry.

By the way, I think that moving the counter into the `CoreWorker` actually would've been necessary in any case. I realized that actor task resubmission can happen concurrently with new task submission, so there would not have been a way to guarantee that the new tasks get sent after the resubmitted ones given the current `CoreWorkerDirectActorSubmitter` interface.

**Review by stephanie-wang — COMMENTED** [review]
This is looking great! One minor suggestion to consider moving the `complete_task_counter_` into the `TaskManager` instead of introducing another callback/dependency, but I don't feel strongly about it since it is nice to have actor state centralized at the `ActorHandle`s.

**Comment by apaszke** [comment]
Hmm yes I'm a bit split on this one too. On one hand it's annoying to have the extra callback (and it can potentially be a bit slower too). On the other hand splitting the actor task state between multiple objects doesn't seem great either.

What about this: the task counters in actor handles are sort of artificial anyway (they are only relevant for handles chosen by the `CoreWorker`), so maybe all of the accounting for actor task counts should be done in the `TaskManager` instead? It's not perfect, because it would introduce some actor-specific fields within the `TaskManager`, but I think it's still an improvement.

**Comment by stephanie-wang** [comment]
> Hmm yes I'm a bit split on this one too. On one hand it's annoying to have the extra callback (and it can potentially be a bit slower too). On the other hand splitting the actor task state between multiple objects doesn't seem great either.
> 
> What about this: the task counters in actor handles are sort of artificial anyway (they are only relevant for handles chosen by the `CoreWorker`), so maybe all of the accounting for actor task counts should be done in the `TaskManager` instead? It's not perfect, because it would introduce some actor-specific fields within the `TaskManager`, but I think it's still an improvement.

That sounds good to me! Long-term, we'll probably want to refactor some of the actor state accounting, but I think it's a good step to at least move actor task counts into the `TaskManager`.

File: src/ray/core_worker/task_manager.h [hunk 11]
```diff
@@ -96,8 +98,21 @@ class TaskManager : public TaskFinisherInterface {
...
 
+  void SetActorCounter(TaskSpecBuilder &builder, const ActorID &actor_id);
+
+  void SetRemainingRetries(const TaskID &task_id, uint64_t retries);
+
   /// Return the number of pending tasks.
-  int NumPendingTasks() const { return pending_tasks_.size(); }
+  int NumPendingTasks() const {
+    absl::MutexLock lock(&mu_);
+    return pending_tasks_.size();
```
Comment (apaszke) [inline src/ray/core_worker/task_manager.h]: This is unrelated to the change, but it looks like this method has been unprotected 🤷‍♂

**Comment by apaszke** [comment]
Ok I think I'm done with the new implementation. It's still missing tests, but should be ready for review.

File: src/ray/core_worker/task_manager.h (hunk 11 above)
Comment (stephanie-wang) [reply src/ray/core_worker/task_manager.h]: Whew thanks!

File: src/ray/core_worker/core_worker.cc [hunk 12]
```diff
@@ -190,13 +194,22 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
...
+        const auto now = current_time_ms();
         absl::MutexLock lock(&mutex_);
-        to_resubmit_.push_back(std::make_pair(current_time_ms() + 5000, spec));
+        if (spec.IsActorTask()) {
+          const auto actor_id = spec.ActorId();
+          // Schedule a kill task unless we already have one in flight.
+          if (!pending_kill_for_actor_.contains(actor_id)) {
+            pending_actor_kills_.emplace_back(now + kActorRestartDeadlineMillis, spec.ActorId());
+            pending_kill_for_actor_[actor_id] = std::prev(pending_actor_kills_.end());
+          }
```
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: I don't think we should be killing the actor during the retry process, is this needed?

**Comment by ericl** [comment]
Just tried it with this python script and this runs into an error:
```
import ray
import os


@ray.remote
class A:
    def pid(self):
        return os.getpid()

    def f(self, x):
        return x

ray.init()
a = A.remote()

for i in range(10):
    print("iteration", ray.get(a.f.remote(i)))
    pid = ray.get(a.pid.remote())
    print("killing", pid)
    os.kill(pid, 9)
```

**Comment by ericl** [comment]
Just tried it out and I think it's working for a bit, but hit a segfault later on
```
import ray
import os


@ray.remote(max_reconstructions=999)
class A:
    def pid(self):
        return os.getpid()

    def f(self, x):
        return x

ray.init()
a = A.remote()

for i in range(10):
    print("iteration", ray.get(a.f.remote(i)))
    pid = ray.get(a.pid.remote())
    print("killing", pid)
    os.kill(pid, 9)
```

```
2020-02-27 14:05:17,567	INFO resource_spec.py:212 -- Starting Ray with 6.64 GiB memory available for workers and up to 3.34 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-02-27 14:05:17,895	INFO services.py:1083 -- View the Ray dashboard at localhost:8265
iteration 0
killing 7265
E0227 14:05:19.210182  7272 task_manager.cc:150] 18446744073709551615 retries left for task 44ee453cd1e8e28345b95b1c0100, attempting to resubmit.
E0227 14:05:19.210319  7272 core_worker.cc:200] Will resubmit task after a 5 second delay: Type=ACTOR_TASK, Language=PYTHON, function_descriptor={type=PythonFunctionDescriptor, module_name=__main__, class_name=A, function_name=f, function_hash=}, task_id=44ee453cd1e8e28345b95b1c0100, job_id=0100, num_args=2, num_returns=2, actor_task_spec={actor_id=45b95b1c0100, actor_caller_id=ffffffffffffffffffffffff0100, actor_counter=2}
2020-02-27 14:05:19,217	WARNING worker.py:1058 -- A worker died or was killed while executing task ffffffffffffffff45b95b1c0100.
iteration 1
killing 7267
E0227 14:05:25.025079  7272 task_manager.cc:150] 18446744073709551615 retries left for task 6f53dca1f451ca9445b95b1c0100, attempting to resubmit.
E0227 14:05:25.025220  7272 core_worker.cc:200] Will resubmit task after a 5 second delay: Type=ACTOR_TASK, Language=PYTHON, function_descriptor={type=PythonFunctionDescriptor, module_name=__main__, class_name=A, function_name=f, function_hash=}, task_id=6f53dca1f451ca9445b95b1c0100, job_id=0100, num_args=2, num_returns=2, actor_task_spec={actor_id=45b95b1c0100, actor_caller_id=ffffffffffffffffffffffff0100, actor_counter=4}
2020-02-27 14:05:25,034	WARNING worker.py:1058 -- A worker died or was killed while executing task ffffffffffffffff45b95b1c0100.
iteration 2
*** Aborted at 1582841130 (unix time) try "date -d @1582841130" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGSEGV (@0x0) received by PID 7224 (TID 0x7faff3f8c740) from PID 0; stack trace: ***
    @     0x7faff37ccf20 (unknown)
    @     0x7faff38223c1 (unknown)
    @     0x7faff38252ed __libc_malloc
    @     0x7faff0e4a258 operator new()
    @     0x7faff12eef3d google::protobuf::internal::ArenaStringPtr::CreateInstanceNoArena()
    @     0x7faff14e901e ray::rpc::Address::MergeFrom()
    @     0x7faff14f5845 ray::rpc::TaskSpec::MergeFrom()
    @     0x7faff138db15 _ZNSt17_Function_handlerIFvvEZN3ray34CoreWorkerDirectActorTaskSubmitter10SubmitTaskENS1_17TaskSpecificationEEUlvE_E9_M_invokeERKSt9_Any_data
    @     0x7faff13928ef ray::LocalDependencyResolver::ResolveDependencies()
    @     0x7faff138ba82 ray::CoreWorkerDirectActorTaskSubmitter::SubmitTask()
    @     0x7faff136b220 ray::CoreWorker::SubmitActorTask()
    @     0x7faff131eace __pyx_pw_3ray_7_raylet_10CoreWorker_39submit_actor_task()
    @           0x509b45 (unknown)
    @           0x50b4a9 _PyEval_EvalFrameDefault
    @           0x507125 (unknown)
    @           0x508fa0 (unknown)
    @           0x50999d (unknown)
    @           0x50c36e _PyEval_EvalFrameDefault
    @           0x507125 (unknown)
    @           0x508fa0 (unknown)
    @           0x50999d (unknown)
    @           0x50b4a9 _PyEval_EvalFrameDefault
    @           0x507125 (unknown)
    @           0x508fa0 (unknown)
    @           0x50999d (unknown)
    @           0x50b4a9 _PyEval_EvalFrameDefault
    @           0x507125 (unknown)
    @           0x508fa0 (unknown)
    @           0x50999d (unknown)
    @           0x50b4a9 _PyEval_EvalFrameDefault
    @           0x507125 (unknown)
    @           0x50a3b3 PyEval_EvalCode
zsh: segmentation fault  python x.py
```

File: src/ray/rpc/worker/core_worker_client.h [hunk 13]
```diff
@@ -167,8 +169,8 @@ class CoreWorkerClient : public std::enable_shared_from_this<CoreWorkerClient>,
   /// \param[in] address Address of the worker server.
   /// \param[in] port Port of the worker server.
   /// \param[in] client_call_manager The `ClientCallManager` used for managing requests.
-  CoreWorkerClient(const rpc::Address &address, ClientCallManager &client_call_manager)
-      : addr_(address), client_call_manager_(client_call_manager) {
+  CoreWorkerClient(const rpc::Address &address, ClientCallManager &client_call_manager, uint64_t sequence_nr_offset = 0)
+      : addr_(address), client_call_manager_(client_call_manager), sequence_nr_offset_(sequence_nr_offset) {
```
Comment (edoakes) [inline src/ray/rpc/worker/core_worker_client.h]: So the idea is the client subtracts the sequence numbers so that the server always thinks they start at 0? I think this should work for now but I'm not sure it's the best idea to have these out of sync on client and server because I could envision it causing bugs/confusion in the future. What do you think about instead including the "base sequence number" in the proto message?

File: src/ray/rpc/worker/core_worker_client.h [hunk 14]
```diff
@@ -279,6 +281,9 @@ class CoreWorkerClient : public std::enable_shared_from_this<CoreWorkerClient>,
   /// The max sequence number we have processed responses for.
   int64_t max_finished_seq_no_ GUARDED_BY(mutex_) = -1;
 
+  /// An offset subtracted from every sequence number before the request is sent.
+  uint64_t sequence_nr_offset_;
```
Comment (ericl) [inline src/ray/rpc/worker/core_worker_client.h]: Could we call this `starting_sequence_number`?

File: src/ray/core_worker/core_worker.cc [hunk 15]
```diff
@@ -190,13 +194,22 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
...
         RAY_LOG(ERROR) << "Will resubmit task after a 5 second delay: "
                        << spec.DebugString();
+        const auto now = current_time_ms();
         absl::MutexLock lock(&mutex_);
-        to_resubmit_.push_back(std::make_pair(current_time_ms() + 5000, spec));
+        if (spec.IsActorTask()) {
+          const auto actor_id = spec.ActorId();
+          // Schedule a kill task unless we already have one in flight.
+          if (!pending_kill_for_actor_.contains(actor_id)) {
+            pending_actor_kills_.emplace_back(now + kActorRestartDeadlineMillis, spec.ActorId());
```
Comment (edoakes) [inline src/ray/core_worker/core_worker.cc]: If I understand this correctly, the idea is to kill an actor whose tasks have failed if we don't get a notification about it being reconstructed within the deadline (currently hardcoded to 3s). Is this to prevent against tasks that fail without crashing the actor?

I'm not sure that we need this at all - can't we just delegate this handling to the raylet (for now, eventually GCS service/owner) and then here wait for the state to transition via GCS notifications? AFAIK the raylet currently guarantees that the actor's state will eventually transition to the correct state (unless the raylet crashes, in which case its actors will be marked dead).

File: src/ray/rpc/worker/core_worker_client.h (hunk 14 above)
Comment (ericl) [reply src/ray/rpc/worker/core_worker_client.h]: Also, we should make sure to subtract this from client_processed_up_to as well (as well as the sequence_number) in the request.

**Review by edoakes — COMMENTED** [review]
The logic moved into the task manager looks great! Not totally sure about the `KillActor` logic - see inline comment.

File: src/ray/core_worker/core_worker.cc (hunk 12 above)
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: The concern here is about whether we should allow clients to kill the actor. I think we should probably just fail all of the pending tasks on this client instead of killing the actor outright (which could end up failing tasks for other clients that did not receive a transient error).

Comment (apaszke) [reply src/ray/core_worker/core_worker.cc]: Yeah I guess the killing might not be necessary in this version of the patch. I can change it to keep trying to resubmit the tasks indefinitely.

File: src/ray/rpc/worker/core_worker_client.h (hunk 14 above)
Comment (apaszke) [reply src/ray/rpc/worker/core_worker_client.h]: Good point about `client_processed_up_to`. I don't mind the name change

[the remark this answers is not in the document]
Comment (apaszke) [reply src/ray/rpc/worker/core_worker_client.h]: I'll go with `starting_sequence_number` that Eric proposed.

File: src/ray/rpc/worker/core_worker_client.h (hunk 13 above)
Comment (apaszke) [reply src/ray/rpc/worker/core_worker_client.h]: I don't have a super strong preference, but this seems slightly better to me. You can avoid sending those few bytes unnecessarily, and it would cause all the confusion to leak into the server as well, making the situation even worse IMO.

File: src/ray/core_worker/core_worker.cc (hunk 15 above)
Comment (apaszke) [reply src/ray/core_worker/core_worker.cc]: Yeah I think we don't need to keep killing the actors anymore. It was important in my previous patch, but now I can just keep resubmitting the tasks indefinitely.

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 16]
```diff
@@ -124,18 +124,9 @@ class CoreWorkerDirectActorTaskSubmitter {
 
   /// Map from actor id to the actor's pending requests. Each actor's requests
   /// are ordered by the task number in the request.
-  absl::flat_hash_map<ActorID, std::map<int64_t, std::unique_ptr<rpc::PushTaskRequest>>>
+  absl::flat_hash_map<ActorID, std::vector<std::unique_ptr<rpc::PushTaskRequest>>>
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]: Hmm why the change to a vector? Wouldn't this cause the vector to no longer be ordered in some cases?

Comment (apaszke) [reply src/ray/core_worker/transport/direct_actor_transport.h]: It does, but there's no reason why the messages need to be submitted in a fixed order. Basically what was happening before was that requests were ordered in both the client and the server. Sorting in the client is unnecessary (gRPC doesn't retain it anyway), so I removed it. And of course this is valid only because sequence numbers are already fixed when `TaskSpecification`s are created (and later only propagated to `PushTaskRequest`s).

File: src/ray/core_worker/core_worker.cc [hunk 17]
```diff
@@ -868,30 +909,36 @@ bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
...
+          if (actor_handle->IsDirectCallActor()) {
+            // TODO: Remove the check for direct calls. We do not reset for the
+            // raylet codepath because it tries to replay all tasks since the
+            // last actor checkpoint.
+            actor_handle->Reset();
+          }
+          direct_actor_submitter_->DisconnectActor(actor_id, /*dead=*/false);
         }
-        direct_actor_submitter_->DisconnectActor(actor_id, false);
+        cancel_kill_task();
```
[the remark this answers is not in the document]
Comment (apaszke) [reply src/ray/core_worker/core_worker.cc]: Sure, no strong preference here I guess.

**Comment by apaszke** [comment]
So, what's the plan for actor restarts? Should I just let the worker keep resubmitting the tasks until they succeed or the actor dies for good while leaving worker management to the raylet?

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 16 above)
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.h]: I think we needed to do this on the client end too because the receiving end will time out if it receives an out-of-order request and doesn't receive the earlier request in time. Usually it's okay, but there are cases where consecutively submitted tasks will get sent at significantly different times because we don't send until the object dependencies have been resolved. @ericl can confirm.

**Comment by stephanie-wang** [comment]
> So, what's the plan for actor restarts? Should I just let the worker keep resubmitting the tasks until they succeed or the actor dies for good while leaving worker management to the raylet?

Yeah I think we can just keep resubmitting. It's a bit ugly, but it should work.

Eventually, we want to move the actor lifetime management into the workers, so whoever submitted the actor creation initially should also manage the actor's lifetime. Then, anyone with a handle can keep submitting as long as the actor's "owner" is still alive. This is similar to what we do with object lifetimes.

**Comment by apaszke** [comment]
Sure, makes sense. I got rid of the killing which is good because it simplifies the code and gets rid of the segfault that @ericl has mentioned (it was due to incorrect STL iterator management on my side).

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 16 above)
Comment (apaszke) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Yes, I've been concerned about that too, but I've decided that in the end the timeout solution is pretty arbitrary too. Ordering client-side can help a bit, but it seems functionally equivalent to allowing longer wait times (unless the distribution of dependency resolution has an extremely long tail).

In fact, why don't we remove that timeout? It was necessary before because we could never retry the tasks, but the current implementation will keep resubmitting any failed requests indefinitely, so it should not be necessary anymore.

Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Removing the timeout sounds good to me, but we need to make sure to properly handle the case where retries are turn off/the task fails `max_retries` times. Assuming it's the behavior we want, need to make sure that the server's sequence number is increased to "skip" the failed task.

File: src/ray/core_worker/task_manager.h [hunk 18]
```diff
@@ -139,6 +156,14 @@ class TaskManager : public TaskFinisherInterface {
   /// Protects below fields.
   mutable absl::Mutex mu_;
 
+  struct ActorTaskState {
```
Comment (stephanie-wang) [inline src/ray/core_worker/task_manager.h]: Could you document the struct? Also, I slightly prefer `num_submitted` over `submitted` to stay consistent with other places where we use `num_*`.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 19]
```diff
@@ -102,10 +96,8 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(const ActorID &actor_id
     // If there are pending requests, treat the pending tasks as failed.
     auto pending_it = pending_requests_.find(actor_id);
     if (pending_it != pending_requests_.end()) {
-      auto head = pending_it->second.begin();
-      while (head != pending_it->second.end()) {
-        auto request = std::move(head->second);
-        head = pending_it->second.erase(head);
+      auto& requests = pending_it->second;
+      for (auto& request: requests) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Why don't we erase from the queue anymore? Couldn't we end up with duplicate tasks in the queue if the `TaskManager` later retries one of these pending tasks?

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 16 above)
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Hmm yeah that's a good point about retries taking care of the issue. I think it should just work even if the actor has `max_reconstructs=0` since the receiving end deduplicates retried tasks.

Also, @edoakes, I think that issue about "skipping" failed tasks is already handled because of the `client_processed_up_to` field.

**Review by stephanie-wang — COMMENTED** [review]
Looks great! I think we can merge it once some Python tests are added.

**Comment by stephanie-wang** [comment]
Hey @apaszke, are you still planning on working on this? If not, we'll reassign this PR to someone else. Thanks!

**Comment by apaszke** [comment]
@stephanie-wang I have a lot of stuff on my plate right now, so feel free to finish this if you want to get it in! Sorry for letting it sit for so long./
